### PR TITLE
[WIP] Compatibility with Swift 3.0 (up to Xcode 8 GM)

### PR DESCRIPTION
**Pull request by nighthawk**
This includes both changes suggested by Xcode's migration tool as well as manual fixing of the issues that Xcode couldn't migrate.

Action list:

- [ ] Double-check `Int+OAuthSwift.swift`
- [ ] Code review
- [ ] Squash commits
- [x] Get tests to work again
- [x] Fix test failures

### CODE REVIEW

**Comment by cambuilt** [comment]
I needed to make additional manual migration changes for Xcode 8, beta 6, but when trying to authenticate to Yelp, I get INVALID_SIGNATURE consistently and despite numerous code reviews. Using the Swift 2.0 version of OAuthSwift with Xcode 7.3 however, it works fine. 

Any ideas?

**Comment by nighthawk** [comment]
@dongri, @rvangraan, can either of you double check `Int+OAuthSwift.swift`. The migrator failed on that file and I'm not sure if I manual migration is correct.

**Comment by nighthawk** [comment]
I'm still stuck on getting the tests to work again as [`Erik`](https://github.com/phimage/Erik) and its dependencies haven't yet been upgraded to Swift 3. @phimage, any advice on that?

**Comment by cambuilt** [comment]
I have given up on OAuthSwift for Swift 3 and using simple URLRequest and URLSession with dataTasks to authenticate to Yelp. I was only using maybe 1% of the capability of  OAuthSwift and was getting tired of fixing the code for every new beta of Xcode 8.

**Comment by phimage** [comment]
@cambuilt swift3 is in beta, xcode is in beta, what do you expect...	
When coding in Swift 3 beta, get accustomed to updating your previous code
You can make a local repo or remote one with your own updates... (a pod update will not erase your change)

@nighthawk I make a swift3.0 branch today. There is some bug in unit test but all compile
Thanks for all your work

**Comment by nighthawk** [comment]
Tests are compiling again and all but `OAuth2SwiftTests.testExpire()` are succeeding.

File: OAuthSwift/OAuthSwiftHTTPRequest.swift [hunk 1]
```diff
@@ -160,17 +163,16 @@ open class OAuthSwiftHTTPRequest: NSObject, URLSessionDelegate, OAuthSwiftReques
                             }
                         }
                     } else {
-                        localizedDescription = OAuthSwiftHTTPRequest.descriptionForHTTPStatus(self.response.statusCode, responseString: String(data: self.responseData as Data, encoding: self.dataEncoding)!)
+                        errorCode = httpResponse.statusCode
```
Comment (nighthawk) [inline OAuthSwift/OAuthSwiftHTTPRequest.swift]: This line fixes `testExpire` as otherwise `-1` would be returned.

**Comment by nighthawk** [comment]
All tests are succeeding now.

File: OAuthSwift/OAuthSwiftHTTPRequest.swift (hunk 1 above)
Comment (phimage) [inline OAuthSwift/OAuthSwiftHTTPRequest.swift]: Yes that was fixed in master branch

**Comment by cambuilt** [comment]
I thought a fork would at least build, but let me know if that should not be expected, since I am relatively new to Github.

**Comment by nighthawk** [comment]
@cambuilt, this pull request is marked as WIP, which stands for Work in Progress.

For you issue, it would be much appreciated if you can provide a test case to `SignTests.swift` with the inputs and the expected signature.

**Comment by cambuilt** [comment]
Not sure how I can do that without giving you my client id and secret, in this case.

**Comment by nighthawk** [comment]
You could replace those with random strings and see if the different versions of OAuthSwift still come up with different signatures.

From what I've noticed in the past (#165 and #168), the issues were with other parts of the input than the client ID and secret.

**Comment by cambuilt** [comment]
Wouldn't the oauth_timestamp have to be identical for both to get a valid result?

**Comment by nighthawk** [comment]
A fixed timestamp is fine for the tests - check out the other tests for comparison.

**Comment by cambuilt** [comment]
Thanks.

**Comment by phimage** [comment]
I think some function renaming is needed to follow the new guidelines by omitting needless words. 
_but can be done after merging into master if conflict appear when doing it_
examples:
- `authorizeWithCallbackURL(_ callbackURL: URL..` -> `authorize(callbackURL: URL`
- `postOAuthAccessTokenWithRequestTokenByCode(_ code: String..` -> 
- `renewAccessTokenWithRefreshToken(_ refreshToken: String` ->

**Comment by s-aska** [comment]
Probably OAuthSwiftTVOS and OAuthSwiftWatchOS requires a convert.

## OAuthSwiftTVOS
> “Use Legacy Swift Language Version” (SWIFT_VERSION) is required to be configured correctly for targets which use Swift. Use the [Edit > Convert > To Current Swift Syntax…] menu to choose a Swift version or use the Build Settings editor to configure the build setting directly.
> warning: no umbrella header found for target 'OAuthSwiftTVOS', module map will not be generated

## OAuthSwiftWatchOS
> “Use Legacy Swift Language Version” (SWIFT_VERSION) is required to be configured correctly for targets which use Swift. Use the [Edit > Convert > To Current Swift Syntax…] menu to choose a Swift version or use the Build Settings editor to configure the build setting directly.
> warning: no umbrella header found for target 'OAuthSwiftWatchOS', module map will not be generated

**Comment by phimage** [comment]
Not a real convert, just open the targets > build settings look for “Use Legacy Swift Language Version”. Changer "Unspecified" to "No"

**Comment by phimage** [comment]
- create swift-version file with 3.0 inside  see #274

old comment
- open the targets > build settings look for “Use Legacy Swift Language Version”. Changer "Unspecified" to "No"
- rename some functions

**Comment by max-potapov** [comment]
@phimage `OAuthSwift:swift3.0` is 17 (!) commits behind the `OAuthSwift:master` and missing some parts of latest release like `OAuthWebViewControllerDelegate` for example.

May I ask you to rebase`swift3.0` branch and this PR to master branch? It will allow to use `OAuthSwift` with applications which was already converted to Swift 3.0 syntax.

**Comment by phimage** [comment]
If @nighthawk can  squash its commits until 3 days I will merge into master branch ( and do some change function naming, etc...).
 If not I will open master branch codes into xcode 8 and do the conversion job (but I want to keep @nighthawk work if possible, but not 19 commits)

**Comment by nighthawk** [comment]
@phimage, I've squashed the commits and opened a new PR: #277. It's still based on the previous swift 3.0 branch as I'm not entirely comfortable with merging in the changes that have seen been made to master.

**Pull request closed without merging**